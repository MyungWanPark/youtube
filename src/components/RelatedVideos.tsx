import React from 'react';

export default function RelatedVideos({ id }: { id: string }) {
    return <div>{id}</div>;
}
