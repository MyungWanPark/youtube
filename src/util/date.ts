import { format, register } from 'timeago.js';
import koLocale from 'timeago.js/lib/lang/ko.js';

register('ko', koLocale);

export function formatAgo(date: string, lang: string = 'en_US') {
    return format(date, lang);
}
