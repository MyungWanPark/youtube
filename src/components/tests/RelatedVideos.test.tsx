import { render, screen, waitForElementToBeRemoved, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { videosData } from '../../test/videoData';
import RelatedVideos from '../RelatedVideos';
import { withAllContext, withRouter } from '../../test/util';
/**
 * Test 사항
 *
 * 1. Related Videos 렌더링 되는지?
 * 2. videos(with parameter) 와 함께 렌더링이 잘 되는지?
 * 3. render loading
 * 4. render error
 *
 */

describe('Related Videos', () => {
    const youtubeClient = {
        channelImageURL: jest.fn(),
        search: jest.fn(),
        popular: jest.fn(),
        relatedVideo: jest.fn(),
    };

    afterEach(() => {
        youtubeClient.relatedVideo.mockReset();
    });

    test('renders correctly', async () => {
        youtubeClient.relatedVideo.mockImplementation(() => videosData);

        const { asFragment } = renderRelatedVideos();
        await waitForElementToBeRemoved(screen.queryByText('is loading..'));
        expect(asFragment()).toMatchSnapshot();
    });

    test('render with related video', async () => {
        youtubeClient.relatedVideo.mockImplementation((id) => videosData);
        const { asFragment } = renderRelatedVideos();

        expect(youtubeClient.relatedVideo).toHaveBeenCalledWith('id');
        const listitems = await screen.findAllByRole('listitem');
        expect(listitems).toHaveLength(videosData.length);
        expect(asFragment()).toMatchSnapshot();
    });

    test('render loading status', () => {
        renderRelatedVideos();

        expect(screen.getByText('is loading..')).toBeInTheDocument();
    });

    test('render error status', async () => {
        youtubeClient.relatedVideo.mockImplementation(() => {
            throw new Error('network error');
        });

        renderRelatedVideos();

        const errMsg = await screen.findByText('network error...');
        expect(errMsg).toBeInTheDocument();
        /*         await waitFor(() => {
            expect(screen.getByText('network error...')).toBeInTheDocument();
        }); */
    });

    function renderRelatedVideos(prop: string = 'id') {
        return render(
            withAllContext(
                withRouter([
                    {
                        element: <RelatedVideos id={prop} />,
                        path: '/',
                    },
                ]),
                youtubeClient
            )
        );
    }
});
