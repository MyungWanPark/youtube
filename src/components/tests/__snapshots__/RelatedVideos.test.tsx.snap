// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Videos render with related video 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex m-2"
    >
      <img
        alt="title1"
        class="w-60 mr-3"
        src="https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title1
        </p>
        <p
          class="text-sm opacity-80"
        >
          channel title1
        </p>
        <p
          class="text-sm opacity-80"
        >
          2 months ago
        </p>
      </div>
    </li>
    <li
      class="flex m-2"
    >
      <img
        alt="title2"
        class="w-60 mr-3"
        src="https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          channel title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          2 months ago
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Related Videos renders correctly 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex m-2"
    >
      <img
        alt="title1"
        class="w-60 mr-3"
        src="https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title1
        </p>
        <p
          class="text-sm opacity-80"
        >
          channel title1
        </p>
        <p
          class="text-sm opacity-80"
        >
          2 months ago
        </p>
      </div>
    </li>
    <li
      class="flex m-2"
    >
      <img
        alt="title2"
        class="w-60 mr-3"
        src="https://i.ytimg.com/vi/8zIf0XvoL9Y/mqdefault.jpg"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          channel title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          2 months ago
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
